File: day06/main.py
import numpy as np


def read_data(filename):
    with open(filename) as file:
        data = list(map(int, file.read().strip().split(",")))

    array = np.zeros(9, dtype=np.int64)
    for k in data:
        array[k] += 1

    return array


def steps(data, n):
    mat = np.zeros((9, 9), dtype=np.int64)
    for i in range(1, 9):
        mat[i-1][i] = 1
    mat[6][0] = 1
    mat[8][0] = 1

    mat = np.linalg.matrix_power(mat, n)
    data = mat @ data

    return np.sum(data)


if __name__ == '__main__':
    data = read_data("input.txt")

    print("Task1:", steps(data, 80))
    print("Task2:", steps(data, 256))

File: day06/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cstdint>

std::vector<int> read_data(std::ifstream& file) {
    int k;
    char delimeter;
    std::vector<int> result;

    while (file >> k >> delimeter) {
        result.push_back(k);
    }
    result.push_back(k);

    return result;
}

int64_t breed(std::vector<int>& data, int days) {
    std::vector<int64_t> cycle(7, 0);
    std::vector<int64_t> newborn(3, 0);

    for (int k=0; k<data.size(); k++) {
        cycle[data[k]]++;
    }

    int pos_cycle = 0;
    int pos_newborn = 0;
    while (days > 0) {
        newborn[(pos_newborn+2)%3] += cycle[pos_cycle];
        cycle[pos_cycle] += newborn[pos_newborn];
        newborn[pos_newborn] = 0;
        pos_cycle = (pos_cycle+1)%7;
        pos_newborn = (pos_newborn+1)%3;
        days--;
    }

    int64_t result = 0;
    for (int k=0; k<cycle.size(); ++k) {
        result += cycle[k];
    }
    for (int k=0; k<newborn.size(); ++k) {
        result += newborn[k];
    }


    return result;
}


int main() {
    std::ifstream file;
    file.open("input.txt");

    std::vector<int> data = read_data(file);

    file.close();

    int64_t res1 = breed(data, 80);
    int64_t res2 = breed(data, 256);

    std::cout << "Task1: " << res1 << std::endl;
    std::cout << "Task2: " << res2 << std::endl;

    return 0;
}
